perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java

```diff
@@ -394,34 +398,84 @@ public boolean isGroupAssigned(PerunSession sess, Group group, Resource resource
 		for(Group g: groups) {
 			getPerunBl().getAttributesManagerBl().checkGroupIsFromTheSameVoLikeResource(perunSession, g, resource);
 
+			// if group is already assigned with active status, we don't want to change its state or other properties
+			if (isGroupAssignedActive(perunSession, g, resource)) {
+				continue;
+			}
+
 			//first we must assign group
 			try {
-				getResourcesManagerImpl().assignGroupToResource(perunSession, g, resource);
-				getResourcesManagerImpl().assignGroupToResourceState(perunSession, g, resource, GroupResourceStatus.PROCESSING);
-				activateGroupResourceAssignment(perunSession, g, resource, async);
+				getResourcesManagerImpl().assignGroupToResource(perunSession, g, resource, autoAssignSubgroups);
+				GroupResourceStatus status = assignInactive ? GroupResourceStatus.INACTIVE : GroupResourceStatus.PROCESSING;
+
+				if (isGroupAssigned(perunSession, g, resource)) {
+					getResourcesManagerImpl().setGroupResourceStatus(perunSession, g, resource, status);
+				} else {
+					getResourcesManagerImpl().assignGroupToResourceState(perunSession, g, resource, status);
+				}
+
+				// if status is PROCESSING, activate group-resource assignment
+				if (!assignInactive) {
+					activateGroupResourceAssignment(perunSession, g, resource, async);
+				}
 			} catch (GroupAlreadyAssignedException e) {
 				// silently skip
 			} catch (GroupNotDefinedOnResourceException ex) {
 				throw new ConsistencyErrorException(ex);
 			}
-		}
 
+			if (autoAssignSubgroups) {
+				GroupsManagerBl groupsManager = getPerunBl().getGroupsManagerBl();
+				List<Group> subgroups;
+
+				// If it is 'members' group, use all groups from corresponding VO
+				if (g.getName().equals(VosManager.MEMBERS_GROUP)) {
+					// if all VO groups are taken, 'autoAssignSubgroups' flag can be set to false
+					subgroups = groupsManager.getAllGroups(perunSession, groupsManager.getVo(perunSession, g));
+					subgroups.removeIf((group -> group.getName().equals(VosManager.MEMBERS_GROUP)));
+				} else {
+					subgroups = groupsManager.getAllSubGroups(perunSession, g);
+				}
+
+				for (Group subgroup : subgroups) {
+					try {
+						getResourcesManagerImpl().assignAutomaticGroupToResource(perunSession, subgroup, resource, g);
+						GroupResourceStatus status = assignInactive ? GroupResourceStatus.INACTIVE : GroupResourceStatus.PROCESSING;
+
+						if (isGroupAssigned(perunSession, subgroup, resource)) {
+							getResourcesManagerImpl().setGroupResourceStatus(perunSession, subgroup, resource, status);
+						} else {
+							getResourcesManagerImpl().assignGroupToResourceState(perunSession, subgroup, resource, status);
+						}
+
+						// if status is PROCESSING, activate group-resource assignment
+						if (!assignInactive) {
+							// automatic subgroups-resource assignment should be async
+							activateGroupResourceAssignment(perunSession, subgroup, resource, true);
```

Comment:
Now the status even for existing ACTIVE subgroup changes to PROCESSING, saves this status to database with `setGroupResourceStatus` and would pass this condition as `isGroupAssigned` returns` false` for PROCESSING assignments, so the group would be activated again. You should probably not set PROCESSING status in case there is already ACTIVE assignment.